compiler/base.ts

```diff
@@ -823,6 +875,60 @@ export class LineCompiler implements InstrVisitor<void>, ExprVisitor<void> {
     this.block.handle(onward);
   }
 
+  visitDefInstr(def: Def): void {
+    this.beginFunction(def);
+    this.block.begin(def, new FunctionBlock());
+  }
+
+  visitShortDefInstr(def: ShortDef): void {
+    this.beginFunction(def);
+    def.body.accept(this);
+    this.emitByte(OpCode.Return);
+    this.endFunction();
+  }
+
+  public beginFunction(def: Def | ShortDef | Lambda): void {
+    const routine = new Routine(
+      RoutineType.Function,
+      this.filename,
+      def.name,
+      def.params.length,
+    );
+
+    this.parents.push(this.routine);
+    this.routine = routine;
+
+    this.scope.begin();
+
```

Comment:
We would declare `this` within the new scope here, knowing that the function would already be on the stack